server/src/controllers/grades.ts

```diff
@@ -0,0 +1,158 @@
+// SPDX-FileCopyrightText: 2023 The Aalto Grades Developers
+//
+// SPDX-License-Identifier: MIT
+
+import { Request, Response } from 'express';
+
+import Attainable from '../database/models/attainable';
+import UserAttainmentGrade from '../database/models/userAttainmentGrade';
+
+import { ApiError } from '../types/error';
+import {
+  CalculationResult,
+  Formula,
+  formulaChecker,
+  FormulaNode,
+  FormulaParams,
+  getFormula,
+  Status
+} from '../types/formulas';
+import { idSchema } from '../types/general';
+import { HttpCode } from '../types/httpCode';
+
+async function calculate(
+  tree: FormulaNode,
+  presetPoints: Map<FormulaNode, number>
+): Promise<CalculationResult> {
+  if (presetPoints.has(tree)) {
+    return { status: Status.Pass, points: presetPoints.get(tree) };
+  }
+  // Not calculated so far, so calculate the points for lower nodes.
+  const subPoints: Array<CalculationResult> =
+    await Promise.all(
+      tree.subFormulaNodes.map(
+        (subTree: FormulaNode) => calculate(subTree, presetPoints)
+      )
+    );
+  // Based on the results from lower nodes, calculate the points for this node.
+  return await tree.validatedFormula(subPoints);
+}
+
+export async function calculateGrades(
+  req: Request,
+  res: Response
+): Promise<void> {
+  const courseId: number = Number(req.params.courseId);
+  const courseInstanceId: number = Number(req.params.instanceId);
+  await idSchema.validate({ id: courseId }, { abortEarly: false });
+  await idSchema.validate({ id: courseInstanceId }, { abortEarly: false });
+
+  const formulaNodesById: Map<number, FormulaNode> = new Map();
+
+  const attainables: Array<{
+    id: number,
+    attainableId: number,
```

Comment:
IIRC this is a Sequelize issue forcing the use of the name "attainableId", but here it would be good if you can somehow specify that this is specifically the parent attainment's ID.